keanu-project/src/main/java/io/improbable/keanu/algorithms/graphtraversal/DifferentiableChecker.java

```diff
@@ -0,0 +1,90 @@
+package io.improbable.keanu.algorithms.graphtraversal;
+
+import io.improbable.keanu.vertices.Vertex;
+import lombok.experimental.UtilityClass;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.Set;
+
+@UtilityClass
+public class DifferentiableChecker {
+
+    /**
+     * Checks whether a set of vertices are differentiable
+     * w.r.t the latents in a graph.
+     *
+     * @param vertices the collection of vertices to check
+     * @return true if the vertices are differentiable w.r.t latents
+     */
+    public boolean isDifferentiable(Collection<Vertex> vertices) {
+        Queue<Vertex> queue = new LinkedList<>(vertices);
+        Set<Vertex> queued = new HashSet<>(vertices);
+        Set<Vertex> constantValueVerticesCache = new HashSet<>();
+
+        while (!queue.isEmpty()) {
+            Vertex visiting = queue.poll();
+
+            if (!visiting.isDifferentiable()) {
+                if (isVertexConstant(visiting, constantValueVerticesCache)) {
+                    continue;
+                } else {
+                    return false;
+                }
+            }
+
+            Collection<Vertex> nextVertices = visiting.getParents();
+            for (Vertex next : nextVertices) {
+                if (!queued.contains(next) && !constantValueVerticesCache.contains(next)) {
+                    queue.offer(next);
+                    queued.add(next);
+                }
+            }
+        }
+        return true;
+    }
+
+    private boolean isVertexConstant(Vertex vertex, Set<Vertex> constantValueVerticesCache) {
```

Comment:
there's an inconsistency to your method names: isVertexConstant vs. isVertexParents*Value*Constant. I'd suggest dropping the word "value" as it doesn't really aid the understandability.